lib/datasource/terraform-provider/index.ts

```diff
@@ -1,119 +1,116 @@
-import * as packageCache from '../../util/cache/package';
-import { Http } from '../../util/http';
+import { cache } from '../../util/cache/package/decorator';
+import { parseUrl } from '../../util/url';
 import * as hashicorpVersioning from '../../versioning/hashicorp';
+import { Datasource } from '../datasource';
 import { getTerraformServiceDiscoveryResult } from '../terraform-module';
 import type { GetReleasesConfig, ReleaseResult } from '../types';
 import type {
   TerraformProvider,
   TerraformProviderReleaseBackend,
 } from './types';
 
-export const id = 'terraform-provider';
-export const customRegistrySupport = true;
-export const defaultRegistryUrls = [
-  'https://registry.terraform.io',
-  'https://releases.hashicorp.com',
-];
-export const defaultVersioning = hashicorpVersioning.id;
-export const registryStrategy = 'hunt';
+export class TerraformProviderDatasource extends Datasource {
+  static readonly id = 'terraform-provider';
 
-const http = new Http(id);
-
-async function queryRegistry(
-  lookupName: string,
-  registryURL: string,
-  repository: string
-): Promise<ReleaseResult> {
-  const serviceDiscovery = await getTerraformServiceDiscoveryResult(
-    registryURL
-  );
-  const backendURL = `${registryURL}${serviceDiscovery['providers.v1']}${repository}`;
-  const res = (await http.getJson<TerraformProvider>(backendURL)).body;
-  const dep: ReleaseResult = {
-    releases: null,
-  };
-  if (res.source) {
-    dep.sourceUrl = res.source;
-  }
-  dep.releases = res.versions.map((version) => ({
-    version,
-  }));
-  // set published date for latest release
-  const latestVersion = dep.releases.find(
-    (release) => res.version === release.version
-  );
-  // istanbul ignore else
-  if (latestVersion) {
-    latestVersion.releaseTimestamp = res.published_at;
+  constructor() {
+    super(TerraformProviderDatasource.id);
   }
-  dep.homepage = `${registryURL}/providers/${repository}`;
-  logger.trace({ dep }, 'dep');
-  return dep;
-}
 
-// TODO: add long term cache (#9590)
-async function queryReleaseBackend(
-  lookupName: string,
-  registryURL: string,
-  repository: string
-): Promise<ReleaseResult> {
-  const backendLookUpName = `terraform-provider-${lookupName}`;
-  const backendURL = registryURL + `/index.json`;
-  const res = (await http.getJson<TerraformProviderReleaseBackend>(backendURL))
-    .body;
+  readonly defaultRegistryUrls = [
+    'https://registry.terraform.io',
+    'https://releases.hashicorp.com',
+  ];
+
+  readonly defaultVersioning = hashicorpVersioning.id;
+
+  readonly registryStrategy = 'hunt';
 
-  if (!res[backendLookUpName]) {
-    return null;
+  @cache({
+    namespace: `datasource-${TerraformProviderDatasource.id}`,
+    key: (getReleasesConfig: GetReleasesConfig) =>
+      `${getReleasesConfig.registryUrl}/${
+        getReleasesConfig.lookupName.includes('/')
+          ? getReleasesConfig.lookupName
+          : `hashicorp/${getReleasesConfig.lookupName}`
+      }`,
+  })
```

Comment:
I'm on the fence about this. On one hand I know it's a bit difficult to read, on the other I'm not sure it's _that_ bad to be worth extracting to a separate function.

Reply:
It's used two times, so i think it's s good candidate for extraction